drivers/net/adin1110/adin1110.c

```diff
@@ -0,0 +1,808 @@
+	int ret;
+
+	val = no_os_field_prep(ADIN1110_MDIO_ST, 0x0);
+	val |= no_os_field_prep(ADIN1110_MDIO_OP, ADIN1110_MDIO_OP_ADDR);
+	val |= no_os_field_prep(ADIN1110_MDIO_PRTAD, phy_id);
+	val |= no_os_field_prep(ADIN1110_MDIO_DEVAD, dev_id);
+	val |= no_os_field_prep(ADIN1110_MDIO_DATA, reg);
+
+	ret = adin1110_reg_write(desc, ADIN1110_MDIOACC(0), val);
+	if (ret)
+		return ret;
+
+	ret = adin1110_wait_mdio_done(desc, 0, &mdio_val);
+	if (ret)
+		return ret;
+
+	val = no_os_field_prep(ADIN1110_MDIO_ST, 0x0);
+	val |= no_os_field_prep(ADIN1110_MDIO_OP, ADIN1110_MDIO_OP_RD);
+	val |= no_os_field_prep(ADIN1110_MDIO_DEVAD, dev_id);
+	val |= no_os_field_prep(ADIN1110_MDIO_PRTAD, phy_id);
+
+	ret = adin1110_reg_write(desc, ADIN1110_MDIOACC(1), val);
+	if (ret)
+		return ret;
+
+	ret = adin1110_wait_mdio_done(desc, 1, &mdio_val);
+	if (ret)
+		return ret;
+
+	*data = no_os_field_get(ADIN1110_MDIO_DATA, mdio_val);
+
+	return 0;
+}
+
+/**
+ * @brief Set a MAC address destination filter, frames who's DA doesn't match
+ * 	  are dropped.
+ * @param desc - the device descriptor
+ * @param mac_address - the MAC filter to be set
+ * @return 0 in case of success, negative error code otherwise
+ */
+static int adin1110_mac_addr_set(struct adin1110_desc *desc,
+				 uint8_t mac_address[ADIN1110_ETH_ALEN])
+{
+	uint32_t reg_val;
+	int ret;
+
+	reg_val = no_os_get_unaligned_be16(&mac_address[0]);
+
+	/* Forward frames from both ports to the host */
+	reg_val |= ADIN1110_MAC_ADDR_APPLY2PORT | ADIN1110_MAC_ADDR_TO_HOST;
+	if (desc->chip_type == ADIN2111)
+		reg_val |= ADIN2111_MAC_ADDR_APPLY2PORT2;
+
+	ret = adin1110_reg_update(desc, ADIN1110_MAC_ADDR_FILTER_UPR_REG,
+				  ADIN1110_MAC_ADDR_FILTER_UPR_MASK, reg_val);
+	if (ret)
+		return ret;
+
+	reg_val = no_os_get_unaligned_be32(&mac_address[2]);
+
+	return adin1110_reg_write(desc, ADIN1110_MAC_ADDR_FILTER_LWR_REG, reg_val);
+}
+
+/**
+ * @brief Write a frame to the TX FIFO.
+ * @param desc - the device descriptor
+ * @param port - the port for the frame to be transmitted on.
+ * @param eth_buff - the frame to be transmitted.
+ * @return 0 in case of success, negative error code otherwise
+ */
+int adin1110_write_fifo(struct adin1110_desc *desc, uint32_t port,
+			struct adin1110_eth_buff *eth_buff)
+{
+	uint32_t field_offset = ADIN1110_WR_HEADER_LEN;
+	uint32_t padding = 0;
+	uint32_t frame_len;
+	int ret;
+
+	struct no_os_spi_msg xfer = {
+		.tx_buff = desc->buff,
+		.cs_change = 1,
+	};
+
+	if (port >= driver_data[desc->chip_type].num_ports)
+		return -EINVAL;
+
+	/*
+	 * The minimum frame length is 64 bytes.
+	 * The MAC is by default configured to add the frame check sequence, so it's
+	 * length shouldn't be added here.
+	 */
+	if (eth_buff->payload_len + ADIN1110_FCS_LEN < 64)
+		padding = 64 - (eth_buff->payload_len + ADIN1110_FCS_LEN);
+
+	frame_len = eth_buff->payload_len + padding + ADIN1110_FRAME_HEADER_LEN;
+
+	/* Align the frame length to 4 bytes */
+	frame_len = frame_len + frame_len % 4;
```

Comment:
This looks incorect. Maybe I don't understand something but if frame_len is in bytes, 9 + 9 % 4 will output 10, not 12 as expected.

`frame_len = (frame_len + 3) & ~0x03` will work as expected by clearing the lowest 2 significant bits and making sure the 3rd bit is a 1 (multiple of 4).